### PR TITLE
Utilizing the reconstructive module in the ServiceWorker

### PR DESCRIPTION
**Pull request by ibnesayeed**
It implements #338 which potentially take care of #335 and #337.

I have added adequate inline documentation for customization. Any generic logic that can go in the reconstructive module should be added there while minimizing customization. There are some known issues in the reconstructive repository which will be fixed soon.

### CODE REVIEW

**Comment by ibnesayeed** [comment]
These are untested changes which should not be merged without proper testing.

File: ipwb/serviceWorker.js [hunk 1]
```diff
@@ -1,76 +1,39 @@
...
-
-function getMyVersion () {
-  fetch(self.location.href)
-  .then(function (resp) {
-    console.log('Running ServiceWorker version ' + resp.headers.get('Server'))
-  })
-}
-
+// This makes a module available named "reconstructive"
+importScripts('https://oduwsdl.github.io/reconstructive/reconstructive.js')
```
Comment (machawk1) [inline ipwb/serviceWorker.js]: This causes the SW to fail in offline testing. Please use the reconstructive local to the replay system.

Comment (ibnesayeed) [reply ipwb/serviceWorker.js]: This has been taken care of. Please test again to see if things are behaving expected.

**Comment by machawk1** [comment]
@ibnesayeed I am encountering an issue (w/ 5b75793dc599e5ff02d075787420db6c0b18ffde) where the replay system is looking for reconstructive.js in the CDXJ, which is causing an infinite reload loop for some reason. This does not occur with serviceWorker.js.

I _think_ this might be resolvable by changing https://github.com/oduwsdl/ipwb/blob/cc80a38ddcacf4482215cd93a344daedba60a432/ipwb/replay.py#L466-L468 to reference reconstructive.js instead of serviceWorker.js but am not currently at a machine where I can test this out.

**Comment by machawk1** [comment]
This also requires a change in ipwb/webui/webui.js. I'm back at an environment where I can test it, so will amend your PR w/ changes to adapt to the new SW script if everything works as expected.

**Comment by machawk1** [comment]
@ibnesayeed I replaced various references of serviceWorker.js to reconstructive.js in dbfd7cfdb1c3e125d6870e9c07b38cbe4dce574a but am still running into an issue where the rerouting is not occurring with reconstructive.

I am running:
`ipwb index ipwb/samples/warcs/froggie.warc.gz | ipwb replay`, navigating to the ipwb homepage where the SW is installed, then navigating to the sole memento. All of this is performed with no Internet connection but the ipfs daemon running.

screen shot 2017-12-18 at 5 59 33 pm

I then repeated this process using cc80a38ddcacf4482215cd93a344daedba60a432 (the current master branch HEAD) and, again being offline, the redirect worked as expected.

screen shot 2017-12-18 at 6 05 19 pm

Something is inconsistent in the rerouting between the old and new SW still that prevents this from being merged just yet.

**Comment by ibnesayeed** [comment]
Changing `serviceWorker.js` references to `reconstructive.js` is not intended and it is not supposed to work, because `reconstructive.js` is not a service worker script, it is a module that can be used in a service worker or offload some logic.

You are encountering this issue, potentially, because you are including the SW registration logic in mementos. This can perhaps be dealt with a custom exclusion.

**Comment by machawk1** [comment]
>  it is a module that can be used in a service worker or offload some logic.

Are we still going to be using `serviceWorker.js`?

**Comment by ibnesayeed** [comment]
> Are we still going to be using `serviceWorker.js`?

Yes, it will be a lightweight script to register the SW and add some customization while offloading main logic to the `reconstructive` module.

I have just pushed a change that reverts your changes and adds a custom exclusion to skip SW handling if the request URL ends with `reconstructive.js`. Hopefully it will solve the recursion issue.

**Comment by machawk1** [comment]
@ibnesayeed That revert also makes it where the serviceWorker is trying to reroute reconstructive.js, which is causing an infinite loop. f158837d5152bf6b5c6e75ed45d22348fecc3a0f

See https://www.youtube.com/watch?v=Fze4s36iJW0&feature=youtu.be

**Comment by ibnesayeed** [comment]
I am trying to understand the dynamics of it. Looking at the `replay.py`, what is the difference between `showMemento` and `showMementoAtDatetime`? I thought mementos were served at `/memento/<datetime>/<urir>` not at `/<datetime>/<urir>`.

**Comment by ibnesayeed** [comment]
I did not realize that you were using an absolute path for the `/serviceWorker.js` file. I have referenced `/reconstructive.js` the same. Hopefully this will get rid of the recursion.

**Comment by machawk1** [comment]
@ibnesayeed Give me a ping when you want me to do a fresh pull and review.

**Comment by ibnesayeed** [comment]
> Give me a ping when you want me to do a fresh pull and review.

Please test it now.

**Comment by machawk1** [comment]
@ibnesayeed The same infinite loop is occurring using 884455a7aa3cc8f8a6f1d6f9ca28faa597b4993f as shown in the screencast above.

**Comment by ibnesayeed** [comment]
Please try to investigate the access log (or by printing request URIs) in the `server.py` file to see where these requests are going and which route are they hitting.

**Comment by machawk1** [comment]
screen shot 2017-12-18 at 7 13 09 pm from #338 ;)

**Comment by ibnesayeed** [comment]
Please check it again.

**Comment by machawk1** [comment]
@ibnesayeed I re-added (in d9dc9b7b60e71e39021efffaab330c5b3e8d8860) recontructive.js to the python setup file so it would be installed and available regardless of if the cloned repo is available.

In testing, the froggie image is still not rewritten when I am offline. See the Python output, JS console, and viewport:
screen shot 2017-12-18 at 7 23 41 pm

**Comment by ibnesayeed** [comment]
> I re-added (in d9dc9b7) recontructive.js to the python setup file so it would be installed and available regardless of if the cloned repo is available.

Thanks! I was going to that, but I was in hurry as I had to go out.

> In testing, the froggie image is still not rerouted when I am offline. See the Python output, JS console, and viewport:

This is expected because of https://github.com/oduwsdl/ipwb/pull/339#issuecomment-352595089. I am still not sure why do we serve mementos at two different addresses. In the JS console, you would see a log entry before the failure like that says `Exclusion found localResource`. Any requests that do not match the `urimPattern` or do not have a referrer that matches that patters would be excluded as a `localResource`.

Unless there is a good reason to serve mementos on two different addresses, we can create a redirect in the `replay` from `/<datetime>/<urir>` to  `/memento/<datetime>/<urir>`.

**Comment by machawk1** [comment]
Re-tested with 3795422ae36395520060b769e23dada7c89cfe0d offline and the frog rerouted! \o/

I do, however, get a 404 for webui.js and an error stating `injectIPWBJS` is not defined, likely associated with the webui inaccessibility.

If I recall, the only reason we request webui.js is to allow the serviceWorker to register from a URI-M if the main ipwb page is not initially hit to register the SW.

There was also a conversation about adding the ability to inject an archive banner for navigation using this injected script.

**Comment by ibnesayeed** [comment]
> Re-tested with 3795422 offline and the frog rerouted! \o/

YAY!

> I do, however, get a 404 for webui.js and an error stating injectIPWBJS is not defined, likely associated with the webui inaccessibility.

I am not really familiar with how you are serving static files. I think the routing logic in the `replay.py` has grown messy and monolithic over the time.

The Reconstrcutive module provides logs to track down any issues, while you can also log requests in in the `replay.py` to see where those requests are hitting.

> If I recall, the only reason we request webui.js is to allow the serviceWorker to register from a URI-M if the main ipwb page is not initially hit to register the SW.

Any request that is routed using the Reconstrcutive module, and not excluded, has a request header (like `X-ServiceWorker: Reconstrcutive:0.1`) in it. If this is missing (or undesired version) in a request that is not a local resource (i.e., it is supposed to be handled by the SW), then the server can redirect user where the appropriate SW will be installed.

> There was also a conversation about adding the ability to inject an archive banner for navigation using this injected script.

There is a placeholder function (which can be overwritten too) in the Reconstrcutive module to specify custom banner. The ability is there, but a default banner is not implemented yet.

**Comment by machawk1** [comment]
I will test with acrts.

In the meantime, I get this error with froggie:

The FetchEvent for "http://matkelly.com/froggies/frog.png" resulted in a network error response: an object that was not a Response was passed to respondWith().

**Comment by machawk1** [comment]
With acrts:

screen shot 2017-12-20 at 5 31 02 pm


localhost_5000_memento_20170206002337_ibnesayeed github io_acrts_ 1

**Comment by ibnesayeed** [comment]
This wrong parameter issue should be fixed now.

**Comment by machawk1** [comment]
The daemons control command exception likely isn't implemented:

screen shot 2017-12-20 at 6 06 28 pm

**Comment by machawk1** [comment]
For the sake of completeness, I also tested acrts:
localhost_5000_memento_20170206002337_ibnesayeed github io_acrts_ 2

but when I killed my network connection, I received the same error as above.

**Comment by ibnesayeed** [comment]
> The daemons control command exception likely isn't implemented.

Those special endpoints are now excluded. The `webui.js` issue should have also been solved with this fix.

**Comment by machawk1** [comment]
Testing w/o an Internet connect, I still get the above daemon issue.

screen shot 2017-12-20 at 9 54 18 pm

screen shot 2017-12-20 at 9 54 15 pm

Your changes also seem to have removed the existence of a host/d14/urir URI-R in favor of a host/memento/d14/urir "scheme".

**Comment by ibnesayeed** [comment]
This JS console suggests that you were running existing SW implementation, not the Reconstructive one. You might want to manually remove the SW and run from fresh with this branch.

No, the `/<datetime>/<urir>` is not removed, it has a server side redirect in place. My analysis on that matter can be seen in #342.

**Comment by machawk1** [comment]
@ibnesayeed You are correct. I've pulled again and re-ran it in Incognito to the same effect. Mind that this is sans Internet connection and the daemon _is_ running.

screen shot 2017-12-21 at 9 33 32 am
screen shot 2017-12-21 at 9 33 33 am
screen shot 2017-12-21 at 9 33 35 am

**Comment by ibnesayeed** [comment]
I can't see a reason why the daemon not running message would be caused by this PR, there might be something else going on. I have pushed some big changes today, you might want to try them again from the clean slate.

**Comment by ibnesayeed** [comment]
Bringing new Docker goodness to this branch.

**Comment by ibnesayeed** [comment]
Steps to perform test on this PR in Docker as of now.

```shell
$ cd /tmp && git clone https://github.com/oduwsdl/ipwb.git && cd ipwb
$ git checkout -b reconstructive-test # Create a test branch from the master
$ git merge origin/issue-338 # Merge reconstructive integration changes in the test branch
$ git merge origin/issue-23 # Merge changes for the outfile support in indexing for easier testing in Docker
$ mkdir -p /tmp/data/{warc,cdxj,ipfs}
$ wget -P /tmp/data/warc/ https://github.com/ibnesayeed/acrts/raw/master/assets/acrts.warc.gz
$ docker image build -t ipwb .
$ docker container run -it --rm -v /tmp/data:/data ipwb ipwb index -o /data/cdxj/acrts.cdxj /data/warc/acrts.warc.gz
$ docker container run -it --rm -v /tmp/data:/data -p 5000:5000 ipwb ipwb replay /data/cdxj/acrts.cdxj
```

Head over to the http://localhost:5000/ then the follow the link to the archived copy of the [ibnesayeed.github.io/acrts/](http://localhost:5000/memento/20170206002337/ibnesayeed.github.io/acrts/) then scroll the page a little and click on the `By JS On Click` block to get all green on the test page!

ipwb-reconstructive-docker-acrts

**Comment by machawk1** [comment]
@ibnesayeed Do you know how reliable it might be to automated this process for use in a CI, e.g., use headless Chrome to scroll the page, take a screenshot, then compare to a reference screenshot for correctness?

**Comment by ibnesayeed** [comment]
It will be little involved to setup a test for the UI, but hopefully it will work. I hope ServiceWorker works in the headless Chrome too.

**Comment by ibnesayeed** [comment]
Just make sure that the window height is set to a smaller number than what is needed for the whole content to fit so that a scroll event can be fired. Also, a click needs to be triggered on one item to make it load the corresponding resource. Also, if the width is set to a smaller number, all the test blocks will be aligned in one column, hence, all the green spots would be vertically aligned too.

After that, we can have a binary pass-fail result based on comparison with a reference image or based on pixel count of specific colors. Alternatively, we can check the color of pixels located in the center of each test indicator and report which test case failed.

**Comment by ibnesayeed** [comment]
I would also note that in the test, we first need to load the homepage so that SW is installed, then follow the archived link from there.

**Comment by ibnesayeed** [comment]
[Puppeteer](https://github.com/GoogleChrome/puppeteer) and [ImageMagick](https://www.imagemagick.org/script/index.php) might be helpful in the process.

**Comment by machawk1** [comment]
@ibnesayeed That is about the process I imagined as well. Thanks for the feedback.

**Comment by ibnesayeed** [comment]
I think we should track this frontend testing in a separate ticket as it is not necessarily tied to this PR.

**Comment by ibnesayeed** [comment]
I have merged new changes from the master branch to simplify testing this in Docker. Now, something like the following should lead to all greens. However, this should also be tested in non-dockerized environment as well as with external network disconnected.

```shell
$ cd /tmp && git clone https://github.com/oduwsdl/ipwb.git && cd ipwb
$ git checkout issue-338
$ mkdir -p /tmp/data/{warc,cdxj,ipfs}
$ wget -P /tmp/data/warc/ https://github.com/ibnesayeed/acrts/raw/master/assets/acrts.warc.gz
$ docker image build -t ipwb .
$ docker container run -it --rm -v /tmp/data:/data ipwb ipwb index -o /data/cdxj/acrts.cdxj /data/warc/acrts.warc.gz
$ docker container run -it --rm -v /tmp/data:/data -p 5000:5000 ipwb ipwb replay /data/cdxj/acrts.cdxj
```

**Comment by machawk1** [comment]
@ibnesayeed Running this procedure I still do not get all-greens from my local system while connected to the Internet.

localhost_5000_memento_20170206002337_ibnesayeed github io_acrts_ 3

**Comment by ibnesayeed** [comment]
These symptoms suggest that the latest Reconstructive is not in action. Could you please clear the cache, unregister any existing SWs, then try again. You may want to share screenshot of browser's console on Slack.

**Comment by ibnesayeed** [comment]
I just performed those steps again and it seems to work for me as expected.

**Comment by machawk1** [comment]
That was the issue. I opened the page in Incognito and got all-green while both online and offline.

**Comment by ibnesayeed** [comment]
Once this PR gets merged, I will create another one to include a rudimentary banner as per #144.

**Comment by ibnesayeed** [comment]
I see some activity in the repo which could potentially make merger of this PR difficult if it remains outstanding for a while.

@machawk1 please proceed with merger, but before that test it again, read the code changes, and let me know if you have any concerns.

**Comment by machawk1** [comment]
Also, note the dependency on fetching a script on GitHub. This fetch needs to be localized as well before we can merge. I'd rather the software not be dependent on GitHub's availability.

**Comment by ibnesayeed** [comment]
> I can't merge this yet, as I have encountered an instance where it breaks in 236bc42.
> Also, note the dependency on fetching a script on GitHub. This fetch needs to be localized as well before we can merge. I'd rather the software not be dependent on GitHub's availability.

All your concerns boil down to just one thing, you were trying to test the very first commit of this PR. There are total of 28 commits to this PR already and the script was added locally on the very same day in the second commit (5b75793).

In order to test this PR you are supposed to checkout the `issue-338` branch, not a specific commit.

**Comment by machawk1** [comment]
Ah, sorry about that, @ibnesayeed. I read the dates wrong. I'll re-test by pulling from the issue-338 branch.

**Comment by ibnesayeed** [comment]
And to see the overall changes in various files after all these commits, please go through [this page](https://github.com/oduwsdl/ipwb/pull/339/files) and add a code review comment if you find something funny.

**Comment by machawk1** [comment]
Re-tested I I get the froggie image while offline without the infinite reload. I'll begin going through the code changes and if all looks well, we should be able to merge.

File: ipwb/reconstructive.js [hunk 2]
```diff
@@ -0,0 +1,371 @@
...
+    let headers = cloneHeaders(event.request.headers);
+    headers.set('X-ServiceWorker', config.id);
+    return new Request(event.request.url, {headers: headers, redirect: 'manual'});
+  }
+
+  /**
+   * cloneHeaders - Clones provided request or response headers.
+   *
+   * @private
+   * @param   {Headers} original - Origina request or response headers.
```
Comment (machawk1) [inline ipwb/reconstructive.js]: "Origina" missing L in comment, not a big issue.

File: ipwb/reconstructive.js [hunk 3]
```diff
@@ -0,0 +1,371 @@
...
+   * @param   {object}     config   - The config object.
+   * @return  {string}              - Potentially modified response.
+   */
+  function createBanner(response, event, config) {
+    // TODO: Add a genric banner markup
+    return '';
+  }
+
+  /**
+   * reroute - The callback fucntion on the fetch event.
```
Comment (machawk1) [inline ipwb/reconstructive.js]: Mis-spelled "fucntion"

File: ipwb/reconstructive.js (hunk 2 above)
Comment (ibnesayeed) [reply ipwb/reconstructive.js]: Will fix it in the upstream Reconstructive repo and bring that change with the upcoming banner related PR.

File: ipwb/reconstructive.js (hunk 3 above)
Comment (ibnesayeed) [reply ipwb/reconstructive.js]: Will be fixed in an upcoming PR.

**Pull request merged**